compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt

```diff
@@ -0,0 +1,113 @@
+                val delegateDescription = if (isGet) "delegate" else "delegate for var (read-write property)"
+                fun reportInapplicableDiagnostics(
+                    candidateApplicability: CandidateApplicability,
+                    candidates: Collection<AbstractFirBasedSymbol<*>>
+                ) {
+                    if (candidateApplicability == CandidateApplicability.INAPPLICABLE_WRONG_RECEIVER) {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_MISSING,
+                            expectedFunctionSignature,
+                            delegateType,
+                            delegateDescription,
+                            context
+                        )
+                    } else {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE,
+                            expectedFunctionSignature,
+                            candidates,
+                            context
+                        )
+                    }
+                }
+                when (val diagnostic = errorNamedReference.diagnostic) {
+                    is ConeUnresolvedNameError -> {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_MISSING,
+                            expectedFunctionSignature,
+                            delegateType,
+                            delegateDescription,
+                            context
+                        )
+                    }
+                    is ConeAmbiguityError -> {
+                        if (diagnostic.applicability.isSuccess) {
+                            // Match is successful but there are too many matches! So we report DELEGATE_SPECIAL_FUNCTION_AMBIGUITY.
+                            reporter.reportOn(
+                                errorNamedReference.source,
+                                FirErrors.DELEGATE_SPECIAL_FUNCTION_AMBIGUITY,
+                                expectedFunctionSignature,
+                                diagnostic.candidates.map { it.symbol },
+                                context
+                            )
+                        } else {
+                            reportInapplicableDiagnostics(diagnostic.applicability, diagnostic.candidates.map { it.symbol })
+                        }
+                    }
+                    is ConeInapplicableCandidateError -> {
+                        reportInapplicableDiagnostics(diagnostic.applicability, listOf(diagnostic.candidate.symbol))
+                    }
+                }
+            }
+        }
+
+        declaration.getter?.acceptChildren(DelegatedPropertyAccessorVisitor(true))
```

Comment:
Maybe safer and slightly faster to just go straight to the body since that's where the delegate calls are anyway, e.g., `declaration.getter?.body?`

Reply:
Good point! Let me do that.